components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/certificatemgt/CertificateManagerImpl.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * you may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.apimgt.impl.certificatemgt;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.apimgt.api.dto.CertificateMetadataDTO;
+import org.wso2.carbon.apimgt.impl.dao.CertificateMgtDAO;
+import org.wso2.carbon.apimgt.impl.utils.CertificateMgtUtils;
+
+import java.io.File;
+import java.util.List;
+
+/**
+ * This class holds the implementation of the CertificateManager interface.
+ */
+public class CertificateManagerImpl implements CertificateManager {
+
+    private static Log log = LogFactory.getLog(CertificateManagerImpl.class);
+    private static final String PROFILE_CONFIG = "sslprofiles.xml";
+    private static String CARBON_HOME = System.getProperty("carbon.home");
+    private static final char SEP = File.separatorChar;
+    private static String SSL_PROFILE_FILE_PATH = CARBON_HOME + SEP + "repository" + SEP + "resources" + SEP
+            + "security" + SEP + PROFILE_CONFIG;
+    private static CertificateMgtDAO certificateMgtDAO = CertificateMgtDAO.getInstance();
+    private static CertificateMgtUtils certificateMgtUtils = new CertificateMgtUtils();
+
+    @Override
+    public ResponseCode addCertificateToPublisher(String certificate, String alias, String endpoint, int tenantId) {
+        try {
+            if (certificateMgtDAO.addCertificate(alias, endpoint, tenantId)) {
+                ResponseCode responseCode = certificateMgtUtils.addCertificateToTrustStore(certificate, alias);
```

Comment:
I hope we saving certificate in the database first and then what is this rollback does? Is first method add cert to database as well before we call the delete certificate in DAO?

Reply:
Ideally if we can save a cert to database but not for gateway we should be indicate it properly. Or rollback accordingly

Reply:
We do not save the certificate in the database. Only the alias and endpoint mapping. 
What I do here is,
1. Persist the Alias|Endpoint mapping.
2. Add to the publisher trust store. 
               If failed to add to publisher ts, then the db entry will also be removed. and operation will return code not equal to success.
3. If adding to publisher ts is successful, then add to the configured gateway environments.